python/cudf/groupby/groupby.py

```diff
@@ -110,6 +112,66 @@ def _apply_basic_agg(self, agg_type, sort_results=False):
         """
         return _cpp_apply_basic_agg(self, agg_type, sort_results=sort_results)
 
+    def apply_multiindex_or_single_index(self, result):
+        if 1 == len(self._by):
+            from cudf.dataframe import index
+            idx = index.as_index(result[self._by[0]])
+            idx.name = self._by[0]
+            result = result.drop(idx.name)
+            if idx.name == self._LEVEL_0_INDEX_NAME:
+                idx.name = None
+            result = result.set_index(idx)
+            return result
+        else:
+            levels = []
+            codes = DataFrame()
+            names = []
+            # Note: This is an O(N^2) solution using gpu masking
+            # to compute new codes for the MultiIndex. There may be
+            # a faster solution that could be executed on gpu at the same
+            # time the groupby is calculated.
+            for by in self._by:
+                level = result[by].unique()
+                code = result[by]
+                for idx, value in enumerate(level):
+                    level_mask = code == value
+                    code = code.masked_assign(idx, level_mask)
+                levels.append(level)
+                codes[by] = code
+                names.append(by)
+            from cudf import MultiIndex
+            multi_index = MultiIndex(levels=levels,
+                                     codes=codes,
+                                     names=names)
+            for by in self._by:
+                result = result.drop(by)
+            return result.set_index(multi_index)
+
+    def apply_multicolumn(self, result, aggs):
+        levels = []
+        codes = []
+        levels.append(self._val_columns)
+        levels.append(aggs)
+        codes.append(list(np.repeat(0, len(aggs))))
```

Comment:
Unsure if the dtype is correct but using zeros should be more efficient and more understandable.
```suggestion
        codes.append(list(np.zeros(len(aggs), dtype='int64')))
```

Reply:
Yeah `zeros` produced 0.0s which gave me pause. I didn't test how to work them or cast them, I just went with this solution instead.